### ISSUE TITLE
Support Pub workspaces

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#5067]
Changes in https://github.com/dart-lang/sdk/issues/53875 might affect assumptions made in the extension here about the location of `package_config.json` relative to a `pubspec.yaml`. When that work is done enough, we should review all places that look for / read a `package_config.json` and ensure they behave correctly in monorepos.

- [x] Initial startup check whether "pub get" needs running
- [ ] TODO: Check anywhere else we're looking at package_configs

**Comment by DanTup** [comment]
https://github.com/dart-lang/sdk/issues/56220#issuecomment-2231373907

**Comment by jakemac53** [comment]
Fwiw, there will still be _a_ pubspec.yaml next to a `.dart_tool/package_config.json`, it will just be in a parent directory, if the pubspec has a `resolution: workspace` entry.

**Comment by jakemac53** [comment]
> 1. If it's `resolution: workspace` then we basically do nothing, except we ensure that the workspace folder (the one with the workspace pubspec - the first one we find with a top-level `workspace:` up the tree) is in the set of projects to check (it may not be if it's out of the workspace).

That sounds right

> 2\. If it's a workspace pubspec (it has a top-level `workspace:`) we look at the pubspec for all of the listed packages and use the oldest pubspec timestamp when doing the check against the package_config, and if it needs running, it runs for the workspace root

s/oldest/newest right? Otherwise that sounds right

**Comment by DanTup** [comment]
> s/oldest/newest right? 

Yep, I was testing you, you passed! 😂

**Comment by DanTup** [comment]
@jakemac53 is it save to run `pub get` in a child workspace folder still? (eg., when a child `pubspec.yaml` is modified, do I need to walk up and run `pub get` in the root, or can I still just run it as if it was for the local project?)

**Comment by DanTup** [comment]
I've made the changes above and this seems to work correctly now with `build`. We won't prompt to run `pub get` because `package_config.json` is missing on the children, and manually editing the pubspecs still triggers `pub get` which Pub handles running for the workspace.

I did find that `pub deps` isn't doing what I'd expect though, so I've filed https://github.com/dart-lang/pub/issues/4330. If it turns out it's working as intended and we need to change how we use it here, I can look at that separately.

I will aim to push these changes to the pre-release extension later today to get them more testing before the release at the end/start of the month.
